pkg/signing/signing.go

```diff
@@ -0,0 +1,130 @@
+package signing
+
+import (
+	"bufio"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"golang.org/x/crypto/sha3"
+)
+
+// Signer allows you to sign a big stream of bytes by calling Sum multiple times, then Sign.
+type Signer struct {
+	state      crypto.KeccakState
+	privateKey *ecdsa.PrivateKey
+}
+
+// HexToECDSA parses a hex-encoded secp256k1 private key string to an ECDSA
+// private key.
+func HexToECDSA(hexKey string) (*ecdsa.PrivateKey, error) {
+	return crypto.HexToECDSA(hexKey)
+}
+
+// FileToECDSA parses a file path to a hex-encoded secp256k1 private key to an
+// ECDSA private key.
+func FileToECDSA(hexPath string) (*ecdsa.PrivateKey, error) {
+	return crypto.LoadECDSA(hexPath)
+}
+
+// NewSigner creates a new signer.
+func NewSigner(pk *ecdsa.PrivateKey) *Signer {
+	return &Signer{
+		state:      sha3.NewLegacyKeccak256().(crypto.KeccakState),
+		privateKey: pk,
+	}
+}
+
+// Sum updates the hash state with a new chunk.
+func (s *Signer) Sum(chunk []byte) {
+	s.state.Write(chunk)
+}
+
+// Sign signs the internal state.
+func (s *Signer) Sign() ([]byte, error) {
+	var h common.Hash
+	_, _ = s.state.Read(h[:])
+	signature, err := crypto.Sign(h.Bytes(), s.privateKey)
+	if err != nil {
+		return []byte{}, fmt.Errorf("sign: %s", err)
+	}
+
+	return signature, nil
+}
+
+// SignFile signs an entire file, returning the signature as a byte slice.
+func (s *Signer) SignFile(filename string) ([]byte, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return []byte{}, fmt.Errorf("error reading [file=%v]: %v", filename, err.Error())
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	// Check if the file is empty and return an error if it is
```

Comment:
this is the only new piece compared to the old impl—since w3s will fail to write an empty file, i figured we should disable trying to upload files with empty content